extension/injections/shared_factory.cpp

```diff
@@ -0,0 +1,131 @@
+//
+// Copyright (c) 2012-2017 Kanstantsin Chernik
+//
+// Distributed under the Boost Software License, Version 1.0.
+// (See accompanying file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+//
+//<-
+#include <cassert>
+#include <memory>
+//->
+#include <boost/di.hpp>
+
+namespace di = boost::di;
+
+template <class T, class TFunc>
+struct shared_factory_impl {
+  shared_factory_impl(TFunc&& creationFunc) : creation_func_(std::move(creationFunc)), is_created_(false) {}
+
+  shared_factory_impl(const shared_factory_impl& other)
+      : creation_func_(std::move(other.creation_func_)), object_(other.object_), is_created_(other.is_created_) {}
+
+  shared_factory_impl(shared_factory_impl&& other) noexcept
+      : creation_func_(std::move(other.creation_func_)), object_(other.object_), is_created_(other.is_created_) {}
+
+  shared_factory_impl& operator=(shared_factory_impl&& other) noexcept {
+    object_ = other.object_;
+    return *this;
+  }
+
+  template <class TInjector, class TDependency>
+  auto operator()(const TInjector& injector, const TDependency&) {
+    if (!is_created_) {
+      auto rebound_injector = di::make_injector(std::move((TInjector&)injector), di::bind<T>().in(di::unique)[di::override]);
+      auto object = creation_func_(rebound_injector);
+      (TInjector&)injector = std::move(rebound_injector);
+
+      //<<should be saved after move otherwise will be overriden>>
+      object_ = object;
+      is_created_ = true;
+    }
+    return object_;
+  }
+
+ private:
+  TFunc&& creation_func_;
+  std::shared_ptr<T> object_;
+  bool is_created_;
+};
+
+template <class T, class TFunc>
+auto shared_factory(TFunc&& creation_func) {
+  return shared_factory_impl<T, TFunc>(std::move(creation_func));
+}
+
+template <class T, class TFunc>
+auto conditional_shared_factory(TFunc condition_func) {
+  return shared_factory<T>([&](const auto& injector) {
+    std::shared_ptr<T> object;
+    if (condition_func()) {
+      object = injector.template create<std::shared_ptr<T>>();
+    }
+    return object;
+  });
+}
+
+//<-
+struct interface {
+  virtual ~interface() noexcept = default;
+  virtual void dummy() = 0;
+};
+struct implementation : interface {
+  static auto& ctor_calls() {
+    static auto calls = 0;
+    return calls;
+  }
+  static auto& dtor_calls() {
+    static auto calls = 0;
+    return calls;
+  }
+  implementation() { ctor_calls()++; }
+  ~implementation() { dtor_calls()++; }
+  void dummy() override {}
+};
+
+//->
+
+int main() {
+  assert(!implementation::ctor_calls());
+  assert(!implementation::dtor_calls());
+
+  {
+    static int calls = 0;
```

Comment:
@krzysztof-jusiak is there any way to make is not static and keep it visible for `conditional_shared_factory`?